docs/design/2020-03-03-SPM-with-parameter-partition.md

```diff
@@ -0,0 +1,109 @@
+select * from t where t.a> = -10 and t.a <= 10;
+```
+
+These two SQLs are the same after standardization, but the data in the table may be concentrated in the [1,10] interval, so the first SQL will be better returned to the table by index, and the second table will be more reasonable.
+
+This is just a simple example. From this example, it can be seen that when there are many predicates, standardized SQL will have many different optimal plans according to the different predicates. Then after the plan evolution is started, it may produce There are a lot of baselines, which will cause the optimizer to build a variety of abstract syntax trees with hints when choosing the baseline to go through the optimization logic, which may become a bottleneck for the optimizer's overhead.On the other hand, too many baselines will make the choice of binding fall back to the choice of CBO, which is also not expected.
+
+## Proposal
+
+### Keys of the problem
+    1. TiDB's current SPM has no elimination mechanism, and it is easy to generate a lot of baselines. It is possible that some baselines are rarely used as the data changes.
+    2. Since we build bind based on standardized SQL, the pros and cons of the plan really depend on the specific value of the predicate.
+
+### Design Overview
+
+Refer to [Adaptive Cursors and SQL Plan Management in Oracle](oracle.com/technical-resources/articles/database/sql-11g-sqlplanmanagement.html), after discussion, we decide to partitioning the bindings by "single column rangeable" predicates.
+
+#### "single column rangeable" predicates
+
+TiDB optimizer uses [Conjunctive Normal Form](https://en.wikipedia.org/wiki/Conjunctive_normal_form) (hereinafter CNF) to divide `where` conditions. For example:
+
+```sql
+where t.a = 1 and 
+      t.b >= 2 and 
+      (t.c*2 == 1 or t.c is null)
+```
+
+Its CNF is $(`t.a=1`) \cap (`t.b>=2`) \cap (`t.c*2==1` \cup `t.c is null`)$. For each single condition, "single column rangeable" is like the format as **${column_name} ${compare_operator} constant**. In this example, the first and second conditions are "single column rangeable".
+
+#### partition the bindings
+
+At first, distinguish two kinds of bindings, *SQL binding* and *Parameter binding*. *SQL binding* has the same behaviour with the origin SQL bind. *Parameter binding* is especially used for SPM plan evolution. Here's a table for differences between them.
+
+binding type   |   Priority    | way to create             | action scope
+---------------|:------------- |:--------------------------|-------------------------------------------------
+SQL bind       | first choice  | use SQL                   | one standardized SQL
+Parameter bind | second choice | enable SPM plan evolution | one standardized SQL with some parameters' range
+
+The parameters' range is the range of the value of the constant which in "single column rangeable" condition.
+
+## Compatibility
+
+Incompatible with older versions of **SQL Bind**.
```

Comment:
We should find a way to make it compatible with SQL Bind in release 3.0. If there is not such method, we should provide an upgrade guide on how to migrate from old SQL Bind to the new SQL Bind.

Reply:
Oh, it's the fault of the document. Actually, it's compatible with SQL Bind in release 3.0.